### DISCUSSION TITLE
Can anyone help compile a .so file for arm64 by editing the cmake files?

### DISCUSSION THREAD

**Discussion by benjaminplong** [ERGO-Code/HiGHS#1600] [General]
I have generated a .so file that works for x86_64 Linux, but now I need an arm version that will also run on Linux. I am not very experienced with CMake so I'm hoping someone here can help me. Thank you.

**Comment by jajhall** [comment]
Could you advise @galabovaa ?

**Comment by mckib2** [comment]
@benjaminplong Are you intending to cross-compile (e.g., build for an ARM target on x86_64) or do you have access to an ARM machine that you can natively compile on?  If you can natively compile, it should "just work" without any need for modification to the cmake build system.  If you are hoping to cross-compile, there could be some easy answers depending on your build environment -- having a description of your build environment (platform, OS, etc.) could be useful

**Comment by galabovaa** [comment]
Additionally, we are working to clean up and improve the CMake build, coming soon with the updated highspy!

**Comment by benjaminplong** [comment]
@mckib2 I am trying to cross-compile from a x86_64 Linux environment to an arm64v8 Linux environment.

**Comment by mckib2** [comment]
@benjaminplong  I got this working on Ubuntu x86_64:

```bash
# install cross-compiler toolchain
sudo apt install gcc-aarch64-linux-gnu g++-aarch64-linux-gnu binutils-aarch64-linux-gnu

# build HiGHS using cross-compiler
mkdir build && cd build
cmake \
    -DCMAKE_C_COMPILER=aarch64-linux-gnu-gcc \
    -DCMAKE_CXX_COMPILER=aarch64-linux-gnu-g++ \
    -DCMAKE_SYSTEM_NAME=Linux \
    -DCMAKE_SYSTEM_PROCESSOR=aarch64 \
    -DFAST_BUILD=OFF \
    -DZLIB=OFF \
    -DSHARED=OFF \
    ..
make -j

# Notes:
# - The following options could go into a toolchain file, but I found it easy enough to add as cmake options:
#     CMAKE_C_COMPILER, CMAKE_CXX_COMPILER, CMAKE_SYSTEM_NAME, CMAKE_SYSTEM_PROCESSOR
# - HiGHS options:
#     FAST_BUILD: turned off to get the option to build static libraries instead of shared
#     ZLIB: cmake was picking up my x86_64 version of the zlib libraries, so I turned it off.  You could probably install the aarch64 versions and it would be fine to leave on
#     SHARED: linker was picking up my x86_64 versions of the std lib, etc., so I changed to building static libraries.  You could probably install the aarch64 versions of the shared libraries and it would be fine to build shared.
```

I tested like this:
```bash
# get an aarch64 emulator
sudo apt install qemu-user qemu-user-static

# run tests using emulator
qemu-aarch64 -L /usr/aarch64-linux-gnu/ bin/unit_tests
```

Curiously I got 1 failure:
```catch.hpp
-------------------------------------------------------------------------------
LP-solver
-------------------------------------------------------------------------------
HiGHS/check/TestLpSolvers.cpp:229
...............................................................................

HiGHS/check/TestLpSolvers.cpp:282: FAILED:
  REQUIRE( info.simplex_iteration_count == 472 )
with expansion:
  437 (0x1b5) == 472 (0x1d8)

===============================================================================
test cases:     186 |     185 passed | 1 failed
assertions: 1253705 | 1253704 passed | 1 failed
```

It's not an important failure it looks like -- just looks like the simplex method took fewer iterations than expected to solve a problem on aarch64 compared to x86_64.

**Comment by benjaminplong** [comment]
![image](https://github.com/ERGO-Code/HiGHS/assets/2343564/284f359d-c7c3-4362-8ec8-260bc4d224fc)
![image](https://github.com/ERGO-Code/HiGHS/assets/2343564/46a65b10-dfa1-4ff8-b5ee-a25ac9ee084c)
@mckib2 I attempted to follow your instructions and have hit the error in the second photo. I am using RockyLinux, I think. The linux install is setup in a docker container by an automated process so I'm not positive what it is, but the package manager is dnf and the package names align with Fedora packages.

Maybe I need the shared libraries you mentioned? I need to produce a .so object for my purposes.

**Comment by mckib2** [comment]
In my experience, CentOS/RHEL distros (of which RockyLinux is trying to be like) are really hard to cross compile on as I believe they don't allow installation of foreign architecture packages (e.g., installing aarch64 packages on x86_64 platforms).  Is there a chance you can switch to a Debian-based distro or is your heart set on RockyLinux?  I'll try to give it a go, but it will likely get really messy and I'm not confident I'd be able to get working binaries

**Reply by benjaminplong** [reply]
@mckib2 My other option would be cross compiling from a windows machine. I think I would then be able to copy the .so to my Linux instance and still use it.

**Comment by mckib2** [comment]
@benjaminplong  Here's my best attempt.  It seems to produce binaries, but I was unable to verify they work as you can't run QEMU inside of Docker.  If you get this building, I'd be curious to know if the libraries work on the target ARM device.

I pulled the latest RockyLinux Docker image and used [`crosstool-NG` ](https://crosstool-ng.github.io/docs) to automatically build the cross-compile toolchain and supporting libraries from source.  All the `yum install`s come from trial and error as I ran into missing tools and libraries, so depending on your install you may run into more or less.  All commands should finish without any error messages, so if you get errors you will need to examine the output and/or logs to see what needs to be fixed.

This method is also able to produce the shared libraries without needing to hunt down any other dependencies as `crosstool-NG` includes the required dependencies for HiGHS by default for the configuration I chose.

```bash
cd /path/to/HiGHS
docker pull rockylinux:9.3
docker run -it -v $(pwd):/src rockylinux:9.3 /bin/bash

$ yum update -y
$ yum install -y epel-release
$ yum repolist
$ dnf config-manager --set-enabled crb
$ yum install -y autoconf automake which flex bzip2 xz unzip help2man file patch libtool bison ncurses-devel gcc gcc-c++ diffutils cmake make texinfo-tex git python3-devel
$ cd /tmp
$ git clone https://github.com/crosstool-ng/crosstool-ng
$ cd crosstool-ng
$ ./bootstrap
$ ./configure --enable-local
$ make -j
$ ./ct-ng aarch64-unknown-linux-gnu  # there are other configurations, run `./ct-ng list-samples` to see all possible options
# may want to customize this futher via `./ct-ng menuconfig` -- I enabled experimental features to run ./ct-ng build as the default Docker root user
# By default, everything was installed into `/root/x-tools/` for me, but it may be dependent on user.  This is configurable via `./ct-ng menuconfig`.
$ ./ct-ng build  # this will take some time to download and build everything

# I needed to run this for some reason -- could be some Docker shenanigans, so you may not need to
$ git config --global --add safe.directory /src

$ cd /src
$ mkdir build && cd build
$ cmake \
    -DCMAKE_C_COMPILER=/root/x-tools/aarch64-unknown-linux-gnu/bin/aarch64-unknown-linux-gnu-gcc \
    -DCMAKE_CXX_COMPILER=/root/x-tools/aarch64-unknown-linux-gnu/bin/aarch64-unknown-linux-gnu-g++ \
    -DCMAKE_SYSTEM_NAME=Linux \
    -DCMAKE_SYSTEM_PROCESSOR=aarch64 \
    ..
$ make -j
```

**Reply by benjaminplong** [reply]
Thank you for all your help @mckib2. I gave up on using the docker image for Rocky Linux because I was running into many missing dependencies and / or PATH issues. Instead I went with your original instructions using WSL + Ubuntu. I was able to run the unit tests just like you and even build the .so object by setting -DShared=ON. Unfortunately the arm environment is still failing to link the .so properly. I'm confident that the .so file is correct for arm so I will need to investigate when the linking isn't working properly. Thanks again for all your help
